File: api_service/api/thread.go
package api

import (
	"db_tp/models"
	"db_tp/services"
	"fmt"
	"github.com/mailru/easyjson"
	"github.com/valyala/fasthttp"
	"strconv"
)

//func CreatePost(ctx *fasthttp.RequestCtx) {
//
//}

func GetThreadDetails(ctx *fasthttp.RequestCtx) {
	slugOrId := ctx.UserValue("slug_or_id").(string)

	thread := services.ThreadSrv.GetBySlugOrId(slugOrId)

	if thread == nil {
		errMsg := models.ErrMsg{Message: fmt.Sprintf("Can't find thread with slug or id:  %s", slugOrId)}
		response, _ := easyjson.Marshal(errMsg)
		ctx.SetBody(response)
		ctx.SetStatusCode(404)
		ctx.SetContentType("application/json")
		return
	}

	resp, _ := easyjson.Marshal(thread)
	ctx.Response.SetBody(resp)
	ctx.SetContentType("application/json")
	ctx.Response.SetStatusCode(200)
}

//func EditThread(ctx *fasthttp.RequestCtx) {
//
//}

func GetThreadPosts(ctx *fasthttp.RequestCtx) {
	slugOrId := ctx.UserValue("slug_or_id").(string)
	limitStr := string(ctx.QueryArgs().Peek("limit"))
	var limit int
	if limitStr == "" {
		limit = 100
	} else {
		limit, _ = strconv.Atoi(limitStr)
	}
	since, _ := strconv.Atoi(string(ctx.QueryArgs().Peek("since")))
	desc, _ := strconv.ParseBool(string(ctx.QueryArgs().Peek("desc")))
	sort := string(ctx.QueryArgs().Peek("sort"))

	var thread *models.Thread
	id, err := strconv.Atoi(slugOrId)
	if err == nil {
		thread = services.ThreadSrv.GetById(id)
	}
	if thread == nil {
		thread = services.ThreadSrv.GetBySlug(slugOrId)
	}
	if thread == nil {
		errMsg := models.ErrMsg{Message: fmt.Sprintf("Can't find posts for thread with slug or id:  %s", slugOrId)}
		response, _ := easyjson.Marshal(errMsg)
		ctx.SetBody(response)
		ctx.SetStatusCode(404)
		ctx.SetContentType("application/json")
		return
	}

	posts := services.ThreadSrv.GetPosts(thread.Id, desc, limit, since, sort)
	resp, _ := easyjson.Marshal(posts)
	ctx.Response.SetBody(resp)
	ctx.SetContentType("application/json")
	ctx.Response.SetStatusCode(200)
}

//func VoteForThread(ctx *fasthttp.RequestCtx) {
//
//}

File: api_service/api/thread.py
from datetime import datetime
from typing import List

from fastapi import APIRouter, Depends

from api.dependency import filter_posts_parameters
from api.exceptions import HttpNotFoundException, HttpConflictException
from db.postgres import foreign_key_violation_exception
from models.post import Post
from models.thread import Thread, ThreadUpdate
from models.vote import Vote
from services.thread import thread_service

router = APIRouter()


@router.post('/{slug_or_id}/create', status_code=201)
async def create_post(
        slug_or_id: str,
        item: List[Post]
) -> List[Post]:
    thread = await thread_service.get_by_slug_or_id(slug_or_id)
    if not thread:
        
        raise HttpNotFoundException()

    if len(item) == 0:
        return item

    now = datetime.now()
    for post in item:
        post.created = now
        post.thread = thread.id
        post.forum = thread.forum

    posts, invalid_parents = await thread_service.create_posts(item)

    
    if invalid_parents:
        raise HttpConflictException()
    if not posts:
        raise HttpNotFoundException()

    return posts


@router.get('/{slug_or_id}/details')
async def get_thread_details(
        slug_or_id: str
) -> Thread:
    thread = await thread_service.get_by_slug_or_id(slug_or_id)

    
    if not thread:
        raise HttpNotFoundException()

    return thread


@router.post('/{slug_or_id}/details')
async def edit_thread(
        slug_or_id: str,
        item: ThreadUpdate
) -> Thread:
    thread = await thread_service.update_by_slug_or_id(slug_or_id, item)
    
    if not thread:
        raise HttpNotFoundException()

    return thread


@router.get('/{slug_or_id}/posts')
async def get_thread_posts(
        slug_or_id: str,
        filter_params: dict = Depends(filter_posts_parameters),
        sort: str = 'flat'
) -> List[Post]:
    thread = None
    if slug_or_id.isdigit():
        thread = await thread_service.get_by_id(int(slug_or_id))

    if not thread:
        thread = await thread_service.get_by_slug(slug_or_id)
    if not thread:
        
        raise HttpNotFoundException()

    posts = await thread_service.get_posts(thread.id, filter_params, sort)

    
    return posts



@router.post('/{slug_or_id}/vote')
async def vote_for_thread(
        slug_or_id: str,
        item: Vote
) -> Thread:
    if slug_or_id.isdigit():
        try:
            await thread_service.vote_by_id(int(slug_or_id), item)
        except foreign_key_violation_exception:
            pass
        else:
            thread = await thread_service.get_by_id(int(slug_or_id))
            return thread

    thread = await thread_service.get_by_slug(slug_or_id)
    if not thread:
        raise HttpNotFoundException()
    try:
        await thread_service.vote_by_id(thread.id, item)
    except foreign_key_violation_exception:
        raise HttpNotFoundException()
    thread = await thread_service.get_by_id(thread.id)

    if not thread:
        raise HttpNotFoundException()

    return thread
